tidy3d/plugins/invdes/region.py

```diff
@@ -0,0 +1,264 @@
+        "``None``, the ``pixel_size`` will determine the FDTD mesh size in the design region. "
+        "Therefore, if your pixel size is large compared to the FDTD grid size, we recommend "
+        "setting the ``override_structure_dl`` directly to "
+        "a value on the same order as the grid size.",
+    )
+
+    transformations: typing.Tuple[FilterProject, ...] = pd.Field(
+        (),
+        title="Transformations",
+        description="Transformations that get applied from first to last on the parameter array."
+        "The end result of the transformations should be the material density of the design region "
+        ". With floating point values between (0, 1), 0 corresponds to the minimum relative "
+        "permittivity and 1 corresponds to the maximum relative permittivity. "
+        "Specific permittivity values given the density array are determined by ``eps_bounds``.",
+    )
+    penalties: typing.Tuple[ErosionDilationPenalty, ...] = pd.Field(
+        (),
+        title="Penalties",
+        description="Set of penalties that get evaluated on the material density. Note that the "
+        "penalties are applied after ``transformations`` are applied. Penalty weights can be set "
+        "inside of the penalties directly through the ``.weight`` field.",
+    )
+
+    override_structure_dl: typing.Union[pd.PositiveFloat, typing.Literal[False]] = pd.Field(
+        None,
+        title="Design Region Override Structure",
+        description="Defines grid size when adding an ``override_structure`` to the "
+        "``JaxSimulation.grid_spec`` corresponding to this design region. "
+        "If left ``None``, ``invdes`` will mesh the simulation with the same resolution as the "
+        "``pixel_size``. "
+        "This is advised if the pixel size is relatively close to the FDTD grid size. "
+        "Supplying ``False`` will completely leave out the override structure.",
+    )
+
+    @staticmethod
+    def _check_params(params: jnp.ndarray):
+        """Ensure ``params`` are between 0 and 1."""
+        if np.any(params < 0) or np.any(params > 1):
+            raise ValueError(
+                "Parameters in the 'invdes' plugin's topology optimization feature "
+                "are restricted to be between 0 and 1."
+            )
+
+    @property
+    def params_shape(self) -> typing.Tuple[int, int, int]:
+        """Shape of the parameters array in (x, y, z), given the ``pixel_size`` and bounds."""
+        # rmin, rmax = np.array(self.geometry.bounds)
+        # lengths = rmax - rmin
+        side_lengths = np.array(self.size)
+        num_pixels = np.ceil(side_lengths / self.pixel_size)
+        # TODO: if the structure is infinite but the simulation is finite, need reduced bounds
+        num_pixels[np.isinf(num_pixels)] = 1
+        return tuple(int(n) for n in num_pixels)
+
+    @property
+    def params_random(self) -> np.ndarray:
```

Comment:
I think it's good to have default initialization helpers. However, by far the most common (and also usually best) way to initialize design regions is uniform 0.5 (due to the projection), and this is also used in the tutorial notebook. Since the invdes plugin is geared toward ease of use, I think constructions like `.params_ones / 2` should be avoided as this requires the user to already have that knowledge (which then renders pre-made initialization a bit pointless). Not really sure how to solve this elegantly though. Maybe a `.params_default` property? And for uniform initialization without noise, maybe instead of a property something like `.params_uniform(x)` where `x` is the value that the array should be initialized to?

Reply:
Hm, the issue with uniform 0.5 params without any randomness added in is that sometimes the symmetry in the design region can lead to a gradient of exactly 0. This is true for the mode converter example, for example. There's essentially 0 conversion from fundamental to higher order mode due to the structure symmetry along y, so you need to break that symmetry initially to get any optimization rolling.

Besides this concern, I'm kind of ok with setting a default param if we can find a universally good one, but I do worry about this edge case as it comes up frequently enough to cause problems.

(I think if we do a default, w just have a property `. params_default` like you suggest and make `params=None` an optional kwarg to `Optimizer.run()` should work fine).

Reply:
Right, that's a thing that happens (I haven't had to deal with it too much). In that case, though, I think the randomness should be as small as possible so there is just enough scattering to lead to a measurable gradient. It shouldn't be in [0, 1] since you run into the problem of having values close to 0 or 1 that are almost like dead pixels in the optimization (depending on the strength of the projection), which then also leads to suboptimal results.

As I'm writing this, I'm wondering if it would make sense to add a check that warns the user if the norm of the gradient is very close to zero after the first adjoint run, and suggest adding some randomness to the initialization? The problem is that this is an edge case that does happen, and it is not intuitive as a beginner to realize what's going on. But adding noise as a default is also not great because it leads to worse performing devices on average vs. uniform initialization (I tested this some time ago).

Reply:
>  I think the randomness should be as small as possible so there is just enough scattering to lead to a measurable gradient

This could work, but if it's too small, (in my experience) the gradient can be extremely close to 0 and the optimizer will have trouble starting.

> I'm wondering if it would make sense to add a check that warns the user if the norm of the gradient is very close to zero after the first adjoint run

I think this could be a good feature to add, yea. Maybe I'll give this a try with initial parameters of 0.5 and then we can set the initial parameters with random variation in our notebook examples, when needed.

Reply:
I changed thins to make `params0` optional in `Optimizer.run()` and use `0.5 * np.ones(...)` as a default.

Also added a warning in the optimizer if the gradient is ever `np.allclose(gradient, 0.0)`

```
                  "All elements of the gradient are almost zero. This likely indicates "
                    "a problem with the optimization set up. This can occur if the symmetry of the "
                    "simulation and design region are preventing any data to be recorded in the "
                    "'output_monitors'. In this case, we recommend creating the initial parameters "
                    " as 'params0 = DesignRegion.params_random' and passing this to "
                    "'Optimizer.run()' to break the symmetry in the design region. "
                    "This zero gradient can also occur if the objective function return value does "
                    "not have a contribution from the input arguments. We recommend carefully "
                    "inspecting your objective function to ensure that the variables passed to the "
                    "function are contributing to the return value."
```